File: analysis/scripts/R4_script_for_plotting_drug_resistance_frequencies_MinION_run_B2.R
############### R script for drug resistance genotype frequencies

### Set wd
path_setwd <- getwd()
setwd(path_setwd)


# Retrieve command-line arguments
args <- commandArgs(trailingOnly = TRUE)

# Define argument names and default values
arg_names <- c("-m", "-g", "-d", "-a")
arg_values <- setNames(rep(NA, length(arg_names)), arg_names)

# Ensure sufficient arguments are provided
if (length(args) < length(arg_names) * 2) {
  stop("\n❌ Missing required arguments.\n\nUsage:\n  Rscript script.R -m MinION_run_name -g genotyping_method -d analysis_date -a analysis_dr\n\nArguments:\n  -m  MinION run name          (e.g. GAMP_22906B2)\n  -g  Genotyping method        (e.g. clair3_diploid)\n  -d  Analysis date            (format: YYYYMMDD, e.g. 20250511)\n  -a  Path to output directory (e.g. /mnt/c/.../output/)\n")
}

# Parse arguments using vectorized lookup
for (i in seq(1, length(args), by = 2)) {
  if (args[i] %in% arg_names) {
    arg_values[args[i]] <- args[i + 1]
  }
}

# Check if any required arguments are missing
if (any(is.na(arg_values))) {
  stop("\n❌ Missing required arguments.\n\nUsage:\n  Rscript script.R -m MinION_run_name -g genotyping_method -d analysis_date -a analysis_dr\n\nArguments:\n  -m  MinION run name          (e.g. GAMP_22906B2)\n  -g  Genotyping method        (e.g. clair3_diploid)\n  -d  Analysis date            (format: YYYYMMDD, e.g. 20250511)\n  -a  Path to output directory (e.g. /mnt/c/.../output/)\n")
}

# Assign parsed values to variables
MinION_run_name <- arg_values["-m"]
genotyping_method <- arg_values["-g"]
analysis_date <- arg_values["-d"]
analysis_dr <- arg_values["-a"]

# Display parsed values
cat("✅ Parsed arguments:\n")
cat("MinION Run Name:", MinION_run_name, "\n")
cat("Genotyping Method:", genotyping_method, "\n")
cat("Analysis Date:", analysis_date, "\n")
cat("Analysis Directory:", paste0(analysis_dr,'/',MinION_run_name,'/',genotyping_method), "\n")


### Load packages
library("tidyverse")
library("readxl")


# Output path
analysis_dr <- paste0(analysis_dr,'/',MinION_run_name,'/',genotyping_method)

### Output filename
plot_dhfr_dhps_haplo_fn <- paste0(analysis_dr, "/plots", "/",MinION_run_name,"_",genotyping_method,"_dhfr_dhps_haplotype_counts_summary_qc_", analysis_date,".png")
plot_GRC_fn <- paste0(analysis_dr, "/plots", "/",MinION_run_name,"_",genotyping_method,"_drug_resistance_counts_summary_qc_", analysis_date,".png")


### Import data
# define paths
run_cov_fn <- paste0(analysis_dr,"/coverage/",MinION_run_name,"_",genotyping_method,"_coverage_by_run_sample",analysis_date,".csv")

# import data
cov_data <- read.csv(run_cov_fn)

## Genotypes for each sample within each MinION run
# define paths
run_geno_fn <- paste0(analysis_dr,"/genotypes/",MinION_run_name,"_genotype_calls_haplotypes_samplerows_v2.csv")

# import data
geno_raw <- read.csv(run_geno_fn)


### Step 1 - Remove all variants below 50x coverage

# Merge the coverage data into the genotype data
geno_cov <- merge(geno_raw, cov_data,
                    by="sample_id",
                    all.x=T,all.y=T)
geno_cov
dim(geno_cov)


### Replace genotypes with NA if coverage too low for that amplicon for that run

## Run
# crt
geno_cov_v2 <- within(geno_cov, {
  crt_227_AC = ifelse(crt_coverage_above_threshold=="FALSE",NA,crt_227_AC)

# dhfr
  dhfr_152_AT = ifelse(dhfr_coverage_above_threshold=="FALSE",NA,dhfr_152_AT)
  dhfr_175_TC = ifelse(dhfr_coverage_above_threshold=="FALSE",NA,dhfr_175_TC)
  dhfr_323_GA = ifelse(dhfr_coverage_above_threshold=="FALSE",NA,dhfr_323_GA)
  dhfr_323_GC = ifelse(dhfr_coverage_above_threshold=="FALSE",NA,dhfr_323_GC)
  dhfr_490_AT = ifelse(dhfr_coverage_above_threshold=="FALSE",NA,dhfr_490_AT)
  dhfr_492_AG = ifelse(dhfr_coverage_above_threshold=="FALSE",NA,dhfr_492_AG)

# dhps
  dhps_1306_TG = ifelse(dhps_coverage_above_threshold=="FALSE",NA,dhps_1306_TG)
  dhps_1307_CT = ifelse(dhps_coverage_above_threshold=="FALSE",NA,dhps_1307_CT)
  dhps_1618_AG = ifelse(dhps_coverage_above_threshold=="FALSE",NA,dhps_1618_AG)
  dhps_1620_AT = ifelse(dhps_coverage_above_threshold=="FALSE",NA,dhps_1620_AT)
  dhps_1742_CG = ifelse(dhps_coverage_above_threshold=="FALSE",NA,dhps_1742_CG)
  dhps_1837_GT = ifelse(dhps_coverage_above_threshold=="FALSE",NA,dhps_1837_GT)
  dhps_1837_GA = ifelse(dhps_coverage_above_threshold=="FALSE",NA,dhps_1837_GA)

# mdr1
  mdr1_256_AT = ifelse(mdr1_coverage_above_threshold=="FALSE",NA,mdr1_256_AT)
  mdr1_551_AT = ifelse(mdr1_coverage_above_threshold=="FALSE",NA,mdr1_551_AT)

# dhfr_haplotype
  dhfr_haplotype = ifelse(dhfr_coverage_above_threshold=="FALSE",NA,dhfr_haplotype)

# dhps_haplotype
  dhps_haplotype = ifelse(dhps_coverage_above_threshold=="FALSE",NA,dhps_haplotype)

# dhps-dhfr haplotype
  dhfr_dhps_haplotype = ifelse((dhfr_coverage_above_threshold=="FALSE" | 
                                 dhps_coverage_above_threshold=="FALSE"),NA,dhfr_dhps_haplotype)

# GRC
  CQ = ifelse(crt_coverage_above_threshold=="FALSE",NA,CQ)
  PYR = ifelse(dhfr_coverage_above_threshold=="FALSE",NA,PYR)
  SX = ifelse(dhps_coverage_above_threshold=="FALSE",NA,SX)
  SP.Rx = ifelse((dhfr_coverage_above_threshold=="FALSE" |
                      dhps_coverage_above_threshold=="FALSE"),NA,SP.Rx)
  SP.IPTp = ifelse((dhfr_coverage_above_threshold=="FALSE" |
                      dhps_coverage_above_threshold=="FALSE"),NA,SP.IPTp)
  ART = ifelse(k13_coverage_above_threshold=="FALSE",NA,ART)
})


n_distinct(geno_cov_v2$sample_id)
  

## Shed all the coverage stats
colnames(geno_cov_v2)
cols <- c("sample_id",
          "crt_227_AC",
          "dhfr_152_AT",
          "dhfr_175_TC",
          "dhfr_323_GA",
          "dhfr_323_GC",
          "dhfr_490_AT",
          "dhfr_492_AG",
          "dhps_1306_TG",
          "dhps_1307_CT",
          "dhps_1618_AG",
          "dhps_1620_AT",
          "dhps_1742_CG",
          "dhps_1837_GT",
          "dhps_1837_GA",
          "mdr1_256_AT",
          "mdr1_551_AT",
          "dhfr_haplotype",
          "dhps_haplotype",
          "dhfr_dhps_haplotype",
          "CQ",
          "PYR",
          "SX",
          "SP.Rx",
          "SP.IPTp",
          "ART")
geno_cov_v3 <- geno_cov_v2 %>% select(all_of(cols))

head(geno_cov_v3)


### Highlight controls
geno_cov_v4 <- within(geno_cov_v3, {
  control <- ifelse((sample_id=="Control_HB3" |
                       sample_id=="Control_Dd2" | 
                       sample_id=="Control_KH2"), "TRUE", "FALSE")
})
geno_cov_v4 %>% count(control)


geno_qc_controls <- geno_cov_v4 %>%
  filter(control=="TRUE")

geno_qc_nocontrols <- geno_cov_v4 %>%
  filter(control=="FALSE")

dim(geno_qc_nocontrols)



############ DATA VIZ

##### Plot 1 - dhfr and dhps genotypes
### Data manipulations

colnames(geno_qc_nocontrols)

geno_qc_nocontrols$crt_227_AC <- as.character(geno_qc_nocontrols$crt_227_AC)
geno_qc_nocontrols$mdr1_256_AT <- as.character(geno_qc_nocontrols$mdr1_256_AT)
geno_qc_nocontrols$mdr1_551_AT <- as.character(geno_qc_nocontrols$mdr1_551_AT)

cols_geno_plot <- c("sample_id",
                    "dhfr_haplotype",
                    "dhps_haplotype",
                    "dhfr_dhps_haplotype")

cols_GRC_plot <- c("sample_id",
                   "CQ",
                   "SX",
                   "PYR",
                   "SP.Rx",
                   "SP.IPTp",
                   "ART")

data_geno_plot <- geno_qc_nocontrols %>%
  select(all_of(cols_geno_plot)) %>%
  rename(DHFR = dhfr_haplotype,
         DHPS = dhps_haplotype,
         DHFR.DHPS = dhfr_dhps_haplotype) %>%
  mutate(DHFR.DHPS = str_replace(DHFR.DHPS, "dhfr-IRNI dhps-AGKAA", "IRNI.AGKAA")) %>%
  mutate(DHFR.DHPS = str_replace(DHFR.DHPS, "dhfr-IRNI dhps-SGEAA", "IRNI.SGEAA")) %>%
  mutate(DHFR.DHPS = str_replace(DHFR.DHPS, "dhfr-IRNI dhps-AAKAA", "IRNI.AAKAA")) %>%
  mutate(DHFR.DHPS = str_replace(DHFR.DHPS, "dhfr-IRNI dhps-SGKAA", "IRNI.SGKAA")) %>%
  mutate(DHFR.DHPS = str_replace(DHFR.DHPS, "dhfr-NRNI dhps-AGKAA", "NRNI.AGKAA")) %>%
  mutate(DHFR.DHPS = str_replace(DHFR.DHPS, "dhfr-NRNI dhps-AAKAA", "NRNI.AAKAA"))

data_GRC_plot <- geno_qc_nocontrols %>%
  select(all_of(cols_GRC_plot))


##### Genotypes

### prepare data for plotting
data_geno_plot2 <- data_geno_plot %>%
  gather(Genotype, Status, -c(sample_id)) %>% # go from wide to long format
  group_by(Genotype) %>%
  count(Status) %>%
  rename(Count = n) %>%
  filter(Genotype != "DHFR.DHPS")
data_geno_plot2 <- drop_na(data_geno_plot2)
head(data_geno_plot2,10)

### Control plot variables
unique(data_geno_plot2$Status)

legend_labels <- c("IRNI", # dhfr triple mutant
                   "NRNI", # dhfr double mutant
                   "NCSI",
                   "AGKAA", # dhps double mutant
                   "SGKAA", # dhps single mutant
                   "AGKAS", # dhps triple mutant
                   "AAKAA")

break_ord <- c("IRNI",
               "NRNI",
               "NCSI",
               "AGKAA",
               "SGKAA",
               "AGKAS",
               "AAKAA")

order <- c("DHFR",
           "DHPS")

### Generate the plots

unique(data_geno_plot2$Status)
n_distinct(data_geno_plot2$Status)

p1_1 <- data_geno_plot2 %>%
  ggplot(aes(x=Genotype, y=Count, fill=Status)) +
  scale_x_discrete(limits = order) +
  geom_bar(stat="identity") +
  theme_minimal() +
  xlab("") +
  ylab("Count") + 
  ylim(0,22) +
  theme(axis.text=element_text(size=10),
        axis.text.x = element_text(angle=30, hjust=1),
        axis.title=element_text(size=11),
        legend.title = element_text(size = 11),
        legend.text = element_text(size = 10),
        panel.background = element_rect(fill = 'white', colour = NA),
        plot.background = element_rect(fill="white", colour = NA)) +
  scale_fill_brewer(palette = "Set2", name = "Genotype", labels = legend_labels, breaks = break_ord)
p1_1


##### Drug resistance predictions
data_GRC_plot2 <- data_GRC_plot %>%
  gather(Drug, Status, -c(sample_id)) %>% # go from wide to long format
  group_by(Drug) %>%
  count(Status) %>%
  rename(Count = n)

data_GRC_plot2 <- drop_na(data_GRC_plot2)

head(data_GRC_plot2, 10)

### Control plot variables
legend_labels_dr <- c("Sensitive",
                      "Resistant")

break_ord_dr <- c("S",
                  "R")

order_dr <- c("CQ",
              "SX",
              "PYR",
              "SP.Rx",
              "SP.IPTp",
              "ART")

cols_dr <- c("#0072B2", # Blue; Sensitive
             "#F0E442") # Yellow; Resistant
# If need an 'Unknown' then use Green, #009E73

p2_1 <- data_GRC_plot2 %>%
  ggplot(aes(x=Drug, y=Count, fill=Status)) +
  scale_x_discrete(limits = order_dr) +
  geom_bar(stat="identity") +
  theme_minimal() +
  xlab("") +
  ylab("Count") + 
  ylim(0,22) +
  theme(axis.text=element_text(size=10),
        axis.text.x = element_text(angle=30, hjust=1),
        axis.title=element_text(size=11),
        legend.title = element_text(size = 11),
        legend.text = element_text(size = 10),
        panel.background = element_rect(fill = 'white', colour = NA),
        plot.background = element_rect(fill="white", colour = NA)) +
  scale_fill_manual(values = cols_dr, name = "Status", labels = legend_labels_dr, breaks = break_ord_dr)
p2_1

p2_1.2 <- data_GRC_plot2 %>%
  ggplot(aes(x=Drug, y=Count, fill=Status)) +
  scale_x_discrete(limits = order_dr) +
  geom_bar(position="fill", stat="identity") +
  theme_minimal() +
  xlab("") +
  ylab("Ratio") + 
  theme(axis.text=element_text(size=10),
        axis.text.x = element_text(angle=30, hjust=1),
        axis.title=element_text(size=11),
        legend.title = element_text(size = 11),
        legend.text = element_text(size = 10) +
  scale_fill_manual(values = cols_dr))
p2_1.2


######### Save outputs
ggsave(plot=p1_1, file=plot_dhfr_dhps_haplo_fn, dpi=200, height=7, width=4, units="in")
ggsave(plot=p2_1, file=plot_GRC_fn, dpi=200, height=4.5, width=7, units="in")

File: analysis/scripts/R4_script_for_plotting_drug_resistance_frequencies_MinION_run_B2.py
import os
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
import argparse

# Argument parser setup
parser = argparse.ArgumentParser(
    description="Process amplicon coverage from NanoRAVE Clair3 diploid output"
)

parser.add_argument("-m", "--minion_run_name", required=True, help="MinION run name (e.g., csp_setA_18aprl2025)")
parser.add_argument("-g", "--genotyping_method", required=True, help="Genotyping method (e.g., clair3_diploid)")
parser.add_argument("-d", "--analysis_date", required=True, help="Analysis date (format: YYYYMMDD)")
parser.add_argument("-a", "--analysis_dr", required=True, help="Path to the output directory")

args = parser.parse_args()

# Assign variables
MinION_run_name = args.minion_run_name
genotyping_method = args.genotyping_method
analysis_date = args.analysis_date
analysis_dr = args.analysis_dr

print("📦 Running coverage script with:")
print(f"   MinION run name     : {MinION_run_name}")
print(f"   Genotyping method   : {genotyping_method}")
print(f"   Analysis date       : {analysis_date}")
print(f"   Output directory    : {analysis_dr}")

# Define paths
analysis_dir = os.path.join(analysis_dr, MinION_run_name, genotyping_method)
plot_dir = os.path.join(analysis_dir, "plots")
os.makedirs(plot_dir, exist_ok=True)

plot_dhfr_dhps_haplo_fn = os.path.join(plot_dir, f"{MinION_run_name}_{genotyping_method}_dhfr_dhps_haplotype_counts_summary_qc_{analysis_date}.png")
plot_GRC_fn = os.path.join(plot_dir, f"{MinION_run_name}_{genotyping_method}_drug_resistance_counts_summary_qc_{analysis_date}.png")

run_cov_fn = os.path.join(analysis_dir, "coverage", f"{MinION_run_name}_{genotyping_method}_coverage_by_run_sample{analysis_date}.csv")
run_geno_fn = os.path.join(analysis_dir, "genotypes", f"{MinION_run_name}_genotype_calls_haplotypes_samplerows_v2.csv")

cov_data = pd.read_csv(run_cov_fn)
geno_raw = pd.read_csv(run_geno_fn)

geno_cov = pd.merge(geno_raw, cov_data, on="sample_id", how="left")

replace_conditions = {
    "crt_227_AC": "crt_227_coverage_above_threshold",
    "dhfr_152_AT": "dhfr_coverage_above_threshold",
    "dhfr_175_TC": "dhfr_coverage_above_threshold",
    "dhfr_323_GA": "dhfr_coverage_above_threshold",
    "dhfr_323_GC": "dhfr_coverage_above_threshold",
    "dhfr_490_AT": "dhfr_coverage_above_threshold",
    "dhfr_492_AG": "dhfr_coverage_above_threshold",
    "dhps_1306_TG": "dhps_coverage_above_threshold",
    "dhps_1307_CT": "dhps_coverage_above_threshold",
    "dhps_1618_AG": "dhps_coverage_above_threshold",
    "dhps_1620_AT": "dhps_coverage_above_threshold",
    "dhps_1742_CG": "dhps_coverage_above_threshold",
    "dhps_1837_GT": "dhps_coverage_above_threshold",
    "dhps_1837_GA": "dhps_coverage_above_threshold",
    "mdr1_256_AT": "mdr1_coverage_above_threshold",
    "mdr1_551_AT": "mdr1_coverage_above_threshold",
    "dhfr_haplotype": "dhfr_coverage_above_threshold",
    "dhps_haplotype": "dhps_coverage_above_threshold",
    "dhfr_dhps_haplotype": ["dhfr_coverage_above_threshold", "dhps_coverage_above_threshold"],
    "CQ": "crt_227_coverage_above_threshold",
    "PYR": "dhfr_coverage_above_threshold",
    "SX": "dhps_coverage_above_threshold",
    "SP.Rx": ["dhfr_coverage_above_threshold", "dhps_coverage_above_threshold"],
    "SP.IPTp": ["dhfr_coverage_above_threshold", "dhps_coverage_above_threshold"],
    "ART": "k13_coverage_above_threshold"
}

for col, cond in replace_conditions.items():
    if isinstance(cond, list):
        if all(c in geno_cov.columns for c in cond):
            mask = ~(geno_cov[cond[0]] == True) | ~(geno_cov[cond[1]] == True)
        else:
            continue  # Skip if any required column is missing
    else:
        if cond in geno_cov.columns:
            mask = ~(geno_cov[cond] == True)
        else:
            continue  # Skip if required column is missing
    geno_cov[col] = geno_cov[col].mask(mask)

cols = list(replace_conditions.keys()) + ["sample_id"]
geno_cov_v3 = geno_cov[cols].copy()


geno_cov_v3["control"] = geno_cov_v3["sample_id"].isin(["Control_HB3", "Control_Dd2", "Control_KH2"])
geno_qc_nocontrols = geno_cov_v3[geno_cov_v3["control"] == False].copy()

geno_qc_nocontrols["crt_227_AC"] = geno_qc_nocontrols["crt_227_AC"].astype(str)
geno_qc_nocontrols["mdr1_256_AT"] = geno_qc_nocontrols["mdr1_256_AT"].astype(str)
geno_qc_nocontrols["mdr1_551_AT"] = geno_qc_nocontrols["mdr1_551_AT"].astype(str)

data_geno_plot = geno_qc_nocontrols[["sample_id", "dhfr_haplotype", "dhps_haplotype", "dhfr_dhps_haplotype"]].rename(columns={
    "dhfr_haplotype": "DHFR",
    "dhps_haplotype": "DHPS",
    "dhfr_dhps_haplotype": "DHFR.DHPS"
})

replacements = {
    "dhfr-IRNI dhps-AGKAA": "IRNI.AGKAA",
    "dhfr-IRNI dhps-SGEAA": "IRNI.SGEAA",
    "dhfr-IRNI dhps-AAKAA": "IRNI.AAKAA",
    "dhfr-IRNI dhps-SGKAA": "IRNI.SGKAA",
    "dhfr-NRNI dhps-AGKAA": "NRNI.AGKAA",
    "dhfr-NRNI dhps-AAKAA": "NRNI.AAKAA"
}
data_geno_plot["DHFR.DHPS"] = data_geno_plot["DHFR.DHPS"].replace(replacements)

data_GRC_plot = geno_qc_nocontrols[["sample_id", "CQ", "SX", "PYR", "SP.Rx", "SP.IPTp", "ART"]]

data_geno_plot2 = data_geno_plot.melt(id_vars="sample_id", var_name="Genotype", value_name="Status")
data_geno_plot2 = data_geno_plot2[data_geno_plot2["Genotype"] != "DHFR.DHPS"]
data_geno_plot2 = data_geno_plot2.dropna().groupby(["Genotype", "Status"]).size().reset_index(name="Count")

order = ["DHFR", "DHPS"]

plt.figure(figsize=(12, 8))  # Large, wide, readable figure
sns.barplot(data=data_geno_plot2, x="Genotype", y="Count", hue="Status", order=order)
plt.ylim(0)
plt.xlabel("\nGenotype", fontsize=16)
plt.ylabel("Sample Count\n", fontsize=16)
plt.xticks(rotation=30, ha="right", fontsize=14)
plt.yticks(fontsize=14)
plt.legend(title="Genotype", loc="center right", bbox_to_anchor=(1.18, 0.5), ncol=1, fontsize=14, title_fontproperties={"weight": "bold"})
sns.despine(top=True, right=True)
plt.tight_layout()
plt.savefig(plot_dhfr_dhps_haplo_fn, dpi=300, bbox_inches="tight")  # High DPI, tight bounding box
print(f'✅ : {plot_dhfr_dhps_haplo_fn}')
plt.close()

data_GRC_plot2 = data_GRC_plot.melt(id_vars="sample_id", var_name="Drug", value_name="Status")
data_GRC_plot2 = data_GRC_plot2.dropna().groupby(["Drug", "Status"]).size().reset_index(name="Count")

order_dr = ["CQ", "SX", "PYR", "SP.Rx", "SP.IPTp", "ART"]
cols_dr = {"S": "#0072B2", "R": "#F0E442"}
legend_labels_dr = ["Sensitive", "Resistant"]
break_ord_dr = ["S", "R"]

plt.figure(figsize=(16, 8))  # Even wider for drug resistance plot
ax = sns.barplot(
    data=data_GRC_plot2,
    x="Drug",
    y="Count",
    hue="Status",
    order=order_dr,
    hue_order=break_ord_dr,
    palette=cols_dr
)
plt.ylim(0)
plt.xlabel("\nDrug", fontsize=16)
plt.ylabel("Sample Count\n", fontsize=16)
plt.xticks(rotation=30, ha="right", fontsize=14)
plt.yticks(fontsize=14)

# Set legend with custom labels, place at center right, slightly outside
handles, labels = ax.get_legend_handles_labels()
ax.legend(
    handles,
    legend_labels_dr,
    title="Resistance Status",
    loc="center right",
    bbox_to_anchor=(1.18, 0.5),
    fontsize=14,
    title_fontproperties={"weight": "bold"}
)
sns.despine(top=True, right=True)
plt.tight_layout()
plt.savefig(plot_GRC_fn, dpi=300, bbox_inches="tight")  # High DPI, tight bounding box
print(f'✅ : {plot_GRC_fn}')
plt.close()


print(f"✅ Plots saved")

